components/threading/thread_viewer/thread_viewer.tsx

```diff
@@ -0,0 +1,564 @@
+        'week',
+        'month',
+        'year',
+    ],
+};
+export function renderView(props: Record<string, any>) {
+    return (
+        <div
+            {...props}
+            className='scrollbar--view'
+        />
+    );
+}
+
+export function renderThumbHorizontal() {
+    return (
+        <div/>
+    );
+}
+
+export function renderThumbVertical(props: Record<string, any>) {
+    return (
+        <div
+            {...props}
+            className='scrollbar--vertical'
+        />
+    );
+}
+
+type Attrs = Pick<HTMLAttributes<HTMLDivElement>, 'className' | 'id'>;
+
+type Props = Attrs & {
+    isCollapsedThreadsEnabled: boolean;
+    userThread?: UserThread | null;
+    posts: Post[];
+    channel: Channel | null;
+    selected: Post | FakePost;
+    previousRhsState?: string;
+    currentUserId: string;
+    currentTeamId: string;
+    previewCollapsed: string;
+    previewEnabled: boolean;
+    socketConnectionStatus: boolean;
+    actions: {
+        removePost: (post: ExtendedPost) => void;
+        selectPostCard: (post: Post) => void;
+        getPostThread: (rootId: string, root?: boolean) => void;
+        getThread: (userId: string, teamId: string, threadId: string, extended: boolean) => unknown;
+        updateThreadRead: (userId: string, teamId: string, threadId: string, timestamp: number) => unknown;
+    };
+    directTeammate: UserProfile;
+    useRelativeTimestamp?: boolean;
+};
+
+type State = {
+    selected?: Post | FakePost;
+    windowWidth?: number;
+    windowHeight?: number;
+    isScrolling: boolean;
+    topRhsPostId: string;
+    openTime: number;
+    postsArray?: Array<Post | FakePost>;
+    isBusy?: boolean;
+    postsContainerHeight: number;
+    userScrolledToBottom: boolean;
+}
+
+export default class ThreadViewer extends React.Component<Props, State> {
+    private scrollStopAction: DelayedAction;
+    private rhspostlistRef: React.RefObject<HTMLDivElement>;
+    private containerRef: React.RefObject<HTMLDivElement>;
+    private postCreateContainerRef: React.RefObject<HTMLDivElement>;
+
+    public static getDerivedStateFromProps(props: Props, state: State) {
+        let updatedState: Partial<State> = {selected: props.selected};
+        if (state.selected && props.selected && state.selected.id !== props.selected.id) {
+            updatedState = {...updatedState, openTime: (new Date()).getTime()};
+        }
+        return updatedState;
+    }
+
+    public constructor(props: Props) {
+        super(props);
+
+        this.scrollStopAction = new DelayedAction(this.handleScrollStop);
+
+        const openTime = (new Date()).getTime();
+
+        this.state = {
+            windowWidth: Utils.windowWidth(),
+            windowHeight: Utils.windowHeight(),
+            isScrolling: false,
+            topRhsPostId: '',
+            openTime,
+            postsContainerHeight: 0,
+            userScrolledToBottom: false,
+        };
+
+        this.rhspostlistRef = React.createRef();
```

Comment:
nit: `rhsPostListRef`?

Reply:
(was moved/existing; didn't want to touch unnecessary)